### ISSUE TITLE
drop python 3.2 and 3.3 compatibility?

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#65]
being compatible with older python versions puts some burden on development:
- tests have to be run on them also (slower test runs, we currently test for 4 python versions)
- some stuff needs special treatment in the code (because it is broken, different or not present)
- we can not use latest releases of some libraries and tools, because they already dropped 3.2 support

python 3.3: https://docs.python.org/3/whatsnew/3.3.html
- new: lzma compression
- new: unicode literals with u"" (this could be interesting if somebody would like to port to 2.7)
- new: getting the terminal dimensions
- new: print(..., flush=True), see sys.stdout.flush() in the code
- new: xattr stdlib code, see #142
- new: mock is in stdlib, see #145
- new: os.replace, .posix_fadvise, .posix_fallocate, .sync, SEEK_HOLE/SEEK_DATA
- new: shutil.disk_usage, stat.filemode, time.process_time
- new: PyMemoryView_FromMemory
- improved: nanosecond precision for stat/utime
- venv comes with it

python 3.4: https://docs.python.org/3/whatsnew/3.4.html
- new: Enum, pathlib, selectors, hashlib.pbkdf2_hmac
- improved: ssl, bytes, bytearray, memoryview, hmac, glob, argparse, os, stat, threading
- pip comes with it

**Comment by ThomasWaldmann** [comment]
note about the tags: "breaking" only in the sense that it breaks 3.2 compatibility (obviously), not in the sense that repo format changes just because of this.

**Comment by maltefiala** [comment]
From a server admin's perspective, I'd say drop it. RHEL / Centos don't ship Python3 anyways. Jessie and Trusty ship 3.4:
* https://packages.debian.org/search?keywords=python3&searchon=names&suite=stable&section=all
* http://packages.ubuntu.com/trusty/python3-minimal

**Comment by RonnyPfannschmidt** [comment]
+10 ^^ im interested in helping with a python2 port (but its a far fetched item wrt time)

**Comment by ThomasWaldmann** [comment]
@RonnyPfannschmidt I also thought about dropping 3.2 because it would make a py2(.7) port way easier.

But I don't really need a py2 port myself, so I think I would not start a port on my own (but I could give advice in case someone wants to do it).

**Comment by ThomasWaldmann** [comment]
Just as an idea:

IF we have good working and compatible binaries that include python 3.4/3.5 and all required libs (except glibc, which needs to come from the installed OS), shall we drop 3.2 and 3.3 early as at least everbody on linux could use the binaries (or install a newer python using backports, pyenv, manually or however).

See reasons (from a developer perspectice) in first post, this is just to get borg user / admin feedback - what do you think?

**Comment by anarcat** [comment]
Python releases inventory:

Distribution         | Version | Notes
-------------------- | ------- | -----
Debian 6.x Squeeze   | 3.1.3   | `oldoldstable`, LTS until 2016-02
Debian 7.x Wheezy    | 3.2.3   | `oldstable`, supported until 2016-02, LTS until 2015-05
Debian 8.x Jessie    | 3.4.2   | `stable`, supported until 2018-05, LTS until 2020-05
Debian 9.x Stretch   | 3.4.3   | `testing`, not yet released
Ubuntu 12.04 Precise | 3.2.3   | `LTS`, supported until 2017-04
Ubuntu 14.04 Trusty  | 3.4.0   | `LTS`, supported until 2019-04
Ubuntu 14.10 Utopic  | 3.4.2   |  unsupported since 2015-07
Ubuntu 15.04 Vivid   | 3.4.3   | supported until 2016-01
Ubuntu 15.10 Wily    | 3.4.3   | to be released in 2015-10-22, supported until 2016-07
CentOS 5/6/7         | nil     | do not ship Python 3, only Python 2.4, 2.6 and 2.7, respectively
Fedora 21/22         | nil     | idem, 2.7 only
RHEL                 | nil     | doesn't ship Python at all
Gentoo               | 3.3, 3.4 | dropped support for 3.2
Arch                 | 3.4 | 3.3 and previous considered "old" (except for 2.7)
Mac OS X             | nil     | 2.7 in 10.8, upstream only for 3.x
Windows              | nil     | does not ship with Python, upstream only

Sources:

 * [Python3-minimal in
   Ubuntu](https://launchpad.net/ubuntu/+source/python3-defaults/)
 * [Python3 in Debian](http://tracker.debian.org/python3)
 * [Redhat derivatives RPM search engine](http://rpm.pbone.net/)
 * [Gentoo](https://wiki.gentoo.org/wiki/Project:Python/Implementations)

In other words, Ubuntu precise LTS, Debian LTS (6.x squeeze) and stable (7.x wheezy) all ship 3.2 or earlier. Redhat derivatives are hopeless and don't seem to ship python3 at all, except for certain versions of Fedora. OSX and windows are equally hopeless, Gentoo and Arch more bleeding edge.

While this is a frustrating situation, i think it matters only insofar that we want to ship distribution packages there. Otherwise, people will *have* to build borg from scratch, and that can easily be done in a virtualenv. or people can use the binary packages.
